File: Solution/Sum of Two Integers.java
/*
Question:
Calculate the sum of two integers a and b, but you are not allowed to use the operator + and -.

Example:
Given a = 1 and b = 2, return 3.

*/




Typical Bit manipulation:

General Idea:
& ==> Carry
^ ==> Addition
Then carry left shift (*2) + addition = result


"&" AND operation, for example, 2 (0010) & 7 (0111) => 2 (0010)

"^" XOR operation, for example, 2 (0010) ^ 7 (0111) => 5 (0101)

"~" NOT operation, for example, ~2(0010) => -3 (1101) what??? Don't get frustrated here. It's called two's complement.

1111 is -1, in two's complement

1110 is -2, which is ~2 + 1, ~0010 => 1101, 1101 + 1 = 1110 => 2

1101 is -3, which is ~3 + 1

so if you want to get a negative number, you can simply do ~x + 1


For this, problem, for example, we have a = 1, b = 3,

In bit representation, a = 0001, b = 0011,

First, we can use "and"("&") operation between a and b to find a carry.

carry = a & b, then carry = 0001

Second, we can use "xor" ("^") operation between a and b to find the different bit, and assign it to a,

Then, we shift carry one position left and assign it to b, b = 0010.

Iterate until there is no carry (or b == 0)



Solution:
public class Solution {
    // Iterative
    public int getSum(int a, int b) {
	    if (a == 0) return b;
	    if (b == 0) return a;

	    while (b != 0) {
		    int carry = a & b;
		    a = a ^ b;
		    b = carry << 1;
	    }
	    return a;
    }
}

Solution 2:






Interesting Follow Up:

1. How to do multiplication without + - * /?

 public int getProduct(int a, int b) {
        if (a==0 || b==0) return 0;
        int result = 0;
        while (b != 0) {
            if ((b & 1) != 0) {
                result = getSum(a,result);
            }
            a <<= 1;
            b >>>= 1;
        }
        return result;
    }
    
2. How to do substraction without + -?
public static int getSubtract(int a, int b) {
		if(a == 0) return (~b + 1); //two's complement.
		if(b == 0) return a;

		int borrow = 0;
		while(b != 0) {
			borrow = (~a) & b;
			a = a ^ b;
			b = borrow << 1;
		}
		return a;
	}

File: Python/Sum of Two Integers.py
Sum of Two Integers
Calculate the sum of two integers a and b, but you are not allowed to use the operator + and -.

Example:
Given a = 1 and b = 2, return 3.


Use Mask to get 32 bit signed int 2,147,483,647 to -2,147,483,648

class Solution(object):
    def getSum(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int
        """
        # 32 bits integer max
        MAX = 0x7FFFFFFF
        # 32 bits interger min
        MIN = 0x80000000
        # mask to get last 32 bits
        mask = 0xFFFFFFFF
        while b != 0:
            # ^ get different bits and & gets double 1s, << moves carry
            a, b = (a ^ b) & mask, ((a & b) << 1) & mask
        # if a is negative, get a's 32 bits complement positive first
        # then get 32-bit positive's Python complement negative
        return a if a <= MAX else ~(a ^ mask)


如何得到一个负数的two complement?假设32bit

bin(-1) = '-0b1'
需要使用mask
mask=0xFFFFFFFF 
-1 & mask = '0b11111111111111111111111111111111' (this is the two complement representation)



2**63-1= 9223372036854775807L
sys.maxint=9223372036854775807




Python has more than 32 bits for integers. You can try to run "print 2 ** 31"and Python would shows the exact number correctly, while other languages like Java would not. Java only recognizes -2 ** 31 to 2 ** 31 - 1.

How does integers presented in Python differ from integers in 32-bit e.g. Java?
From what I heard, Python has 64 bits. (Please let me know if I am wrong. )
So 1 in Python would look like 0x0000000000000001, but it looks like 0x00000001 in 32-bit format.
-1 in Python would look like 0xFFFFFFFFFFFFFFFF, but it looks like 0xFFFFFFFF in 32-bit format.

-2 in python is 0xFFFFFFFFFFFFFFFE , but in java it is 0xFFFFFFFE 

-1 + 1 =0x00000001 + 0xFFFFFFFF  = 0x(10000.....0) => 所以还需要一个mask


It seems that the input given by LC is in 32-bit format. Since Python would treat it as positive with 1 on the 32 position, we have to use mask to treat it as negative.